libsweep/src/arch/win/serial_win.cc

```diff
@@ -0,0 +1,280 @@
+
+  // set the timeouts
+  if (!SetCommTimeouts(hComm, &timeouts)) {
+    *error = error_construct("setting serial port timeouts failed");
+    CloseHandle(hComm);
+    return nullptr;
+  }
+
+  // create the overlapped os reader
+  OVERLAPPED osReader = {0};
+  OVERLAPPED osWrite = {0};
+
+  // Create the overlapped read event. Must be closed before exiting
+  osReader.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (osReader.hEvent == NULL) {
+    *error = error_construct("creating overlapped read event failed");
+    CloseHandle(hComm);
+    return nullptr;
+  }
+  // Create the overlapped write event. Must be closed before exiting
+  osWrite.hEvent = CreateEvent(NULL, TRUE, FALSE, NULL);
+  if (osWrite.hEvent == NULL) {
+    *error = error_construct("creating overlapped write event failed");
+    CloseHandle(hComm);
+    return nullptr;
+  }
+
+  // set the comm mask
+  if (!SetCommMask(hComm, EV_RXCHAR | EV_ERR)) {
+    *error = error_construct("setting comm mask failed");
+    CloseHandle(hComm);
+    CloseHandle(osReader.hEvent);
+    CloseHandle(osWrite.hEvent);
+    return nullptr;
+  }
+
+  // purge the comm port of any pre-existing data or errors
+  if (!PurgeComm(hComm, PURGE_RXABORT | PURGE_TXABORT | PURGE_RXCLEAR | PURGE_TXCLEAR)) {
+    *error = error_construct("flushing serial port failed during serial device construction");
+    CloseHandle(hComm);
+    CloseHandle(osReader.hEvent);
+    CloseHandle(osWrite.hEvent);
+    return nullptr;
+  }
+
+  // create the serial device
+  auto out = new device{hComm, osReader, osWrite};
+
+  return out;
+}
+
+void device_destruct(device_s serial) {
+  SWEEP_ASSERT(serial);
+
+  error_s ignore = nullptr;
+  device_flush(serial, &ignore);
+
+  // try to close the serial port
+  if (!CloseHandle(serial->fd)) {
+    ignore = error_construct("closing serial port failed");
+  }
+
+  // try to close the overlapped read event
+  if (!CloseHandle(serial->ro_.hEvent)) {
+    ignore = error_construct("closing serial port overlapped read event failed");
+  }
+
+  // try to close the overlapped write event
+  if (!CloseHandle(serial->wo_.hEvent)) {
+    ignore = error_construct("closing serial port overlapped write event failed");
+  }
+
+  (void)ignore; // nothing we can do here
+
+  delete serial;
+}
+
+void device_read(device_s serial, void* to, int32_t len, error_s* error) {
+  SWEEP_ASSERT(serial);
+  SWEEP_ASSERT(to);
+  SWEEP_ASSERT(len >= 0);
+  SWEEP_ASSERT(error);
+
+  DWORD rx_read = 0;
+  DWORD totalNumBytesRead = 0;
+
+  // read bytes until "len" bytes have been read
+  while (totalNumBytesRead < (DWORD)len) {
+
+    if (!ReadFile(serial->fd, (char*)to + totalNumBytesRead, len - totalNumBytesRead, &rx_read, &(serial->ro_))) {
+      *error = error_construct("reading from serial device failed");
+      return;
+    }
+    totalNumBytesRead += rx_read;
+  }
+
+  SWEEP_ASSERT(totalNumBytesRead == (DWORD)len && "reliable read failed to read requested number of bytes");
```

Comment:
Hm we should error here and not only assert. The callsite should not continue when we fail to read all requested bytes.